Fix null Timestamp/Date coercion to 0 epoch.

aa8y opened the pull request:
The current implementation of `Long` to `Date` / `Timestamp` coercion does not take into account `null` values. Due to this when a `null` is cast into a `Long`, it ends up as 0 (as Scala Long does not support `null`). This PR is a fix for that.

aa8y commented:
@gengliangwang Please let me know if you need anything from me to merge this PR.

aa8y commented:
@JoshRosen and @marmbrus Anything I can do to help move this forward?

gengliangwang commented on src/main/scala/com/databricks/spark/avro/SchemaConverters.scala:
```diff
@@ -154,9 +154,13 @@ object SchemaConverters {
           identity
         case (TimestampType, LONG) =>
-          (item: AnyRef) => new Timestamp(item.asInstanceOf[Long])
+          (item: AnyRef) => Option(item).map { o =>
```
For better performance,  can you follow the code below
```
if (item == null) {
  null
} else {
...
}
```

gengliangwang commented on src/test/scala/com/databricks/spark/avro/AvroSuite.scala:
```diff
@@ -534,6 +534,28 @@ class AvroSuite extends FunSuite with BeforeAndAfterAll {
+      val nullTime: Timestamp = null
+      val nullDate: Date = null
+      val schema = StructType(Seq(
+        StructField("_1", DateType, false), 
```
the field `nullable` should be true here

aa8y commented:
@gengliangwang I've made the requested changes. Please let me know if you need anything else.

gengliangwang commented:
There is a lot `if (item == null)...` in current code, can you remove the redundant code, and create a wrapper over `createConverter` for checking the null if schema is nullable?

I prefer to do it in this PR. If you don't have time, I can merge this one and do it later.

aa8y commented:
@gengliangwang I've made some changes. Let me know if it looks like what you had in mind.

gengliangwang commented on src/main/scala/com/databricks/spark/avro/SchemaConverters.scala:
```diff
@@ -262,37 +239,29 @@ object SchemaConverters {
-                  (item: AnyRef) => if (item == null) {
-                    null
-                  } else {
+                  (item: AnyRef) =>
```
brace

aa8y commented:
@gengliangwang I removed the braces because they were inconsistent within the function. Some places had it and some places didn't. So I removed them from a few places for consistency. I think it reads better this way but if you insist I can add them back.

gengliangwang commented:
@aa8y I see, please keep it.

aa8y commented:
@gengliangwang Is this ready to be merged then?

gengliangwang commented:
@aa8y there are still two comments need to be addressed.

aa8y commented on src/main/scala/com/databricks/spark/avro/SchemaConverters.scala:
```diff
@@ -184,8 +174,11 @@ object SchemaConverters {
-              val converter = createConverter(avroField.schema(), sqlField.dataType,
-                path :+ sqlField.name)
+              val converter = (item: AnyRef) => {
+                if (item == null) item else {
```
I'll fix this.

aa8y commented on src/main/scala/com/databricks/spark/avro/SchemaConverters.scala:
```diff
@@ -198,59 +191,43 @@ object SchemaConverters {
           }
 
-          (item: AnyRef) => {
-            if (item == null) {
```
Wasn't that the whole reason why you wanted add a wrapper around `createConverter()`? Why do we want that `null` check in there if we're checking for it in the wrapper?

gengliangwang replied:
If the schema contains union of [null, record], then we need this code, right?

aa8y replied:
I understand you're good with not changing this then.

aa8y commented:
@gengliangwang Sorry. I didn't see those. I left responses to the comments. I'll wait for your response before I push something.

aa8y commented:
@gengliangwang Sorry for the delay. I pushed a commit to fix the braces for the if-else block.

gengliangwang commented:
Thanks, merge to master
